teps/0015-pending-pipeline.md

```diff
@@ -0,0 +1,121 @@
+authors:
+  - "@jbarrick-mesosphere"
+creation-date: 2020-09-10
+last-updated: 2020-09-10
+status: implementable
+---
+
+# TEP-0015: Support Pending PipelineRuns
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+    - [Platform Implementor](#platform-implementor)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+This TEP proposes a method for pausing `PipelineRuns` that are then ignored by
+the Tekton reconciler. These pending resources can be used by platform
+implementers to control when a `PipelineRun` can create new `TaskRuns`, for
+example, in cases where the cluster is under heavy load and not ready to process
+new `TaskRuns`.
+
+`PipelineRuns` can be created in a pending state so that they are not started at all.
+
+## Motivation
+
+The primary motivation for this is to support operators who need to control the
+start up of `PipelineRuns` in clusters that are under heavy load. By allowing
+`PipelineRuns` to be created without actually starting them, one could implement
+a custom controller to implement any `PipelineRun` scheduling policy that is
+transparent to the `PipelineRun` controller and `PipelineRun` creator (e.g.,
+Tekton triggers).
+
+With this feature, platforms can control when `PipelineRuns` are started without
+introducing a bespoke system for storing `PipelineRuns` - potentially adding new
+failure scenarios or UI complexity.
+
+### Goals
+
+* Provide a pending mechanism for `PipelineRuns` that operators can use
+  to control whether or not a `PipelineRun` is running (e.g., "creating new
+  `TaskRuns`").
+
+### Non-Goals
+
+* Support pausing of a running `TaskRun` as that would requiring pausing the
+  Kubernetes pods themselves.
+
+## Requirements
+
+* Provide a mechanism to operators that can be used to mark a `PipelineRun` in a pending
+  state.
+* `PipelineRuns` that are created with the pending setting enabled are not started
+  until the setting is toggled off.
+
+## Proposal
+
+The `PipelineRunSpec` supports a field called `Status` of the type
+`PipelineRunSpecStatus`. Currently the field only supports a single setting,
+`PipelineRunCancelled`. This field will be expanded to support an additional
+setting, `PipelineRunPending`.
+
+When the `Status` field is set to `PipelineRunPending`, the Tekton reconciler
+will set the `Succeeded` `Condition` `Reason` in the `PipelineRun` `Status` to
+`PipelineRunPending`:
+
+```
+status:
+  conditions:
+  - lastTransitionTime: "2020-09-15T18:15:29Z"
+    message: 'PipelineRun is pending.'
+    reason: PipelineRunPending
+    status: "Unknown"
+    type: Succeeded
+```
+
+To start a PipelineRun, the operator should set the `Status` field to empty,
+the Tekton controller will see the change and proceed with the Pipeline as
+normal. Upon starting, the `Succeeded` `Condition` `Reason` will be changed to
```

Comment:
Can we add more details what "normal" means here?

What is reconciler doing after resuming back? Is there any specific validation `reconciler` needs to implement/check to resume back?

`Cancellation` is an easy job, cancel task, update status, and be done. But I am guessing `resuming` is not. What are the possible pipelineRun states which can allow `pipelineRun` going in `pending` mode?

Reply:
Alright, I have tried to clarify this in the document, let me know if it helps.

Reply:
thanks Justin, its not clear to me but thats fine, its just me, please squash the commits before we can merge. We can discuss further in the implementation PR.

I am having hard time understanding, when can I send a `pipelineRun` in pending mode? As soon as a `pipelineRun` is instantiated, it initializes the `pr.Status.Conditions` with `status` set to `unknown` and reason to `started` with the start time. Can a `started` pipeline be assigned `pending` state?

I am assuming here the answer is no, since the pipeline is already marked as started and have started validating resources.

After all the validation is done, `pr.Status` is set to `running` from `started`. How is `webhook` validating that the pipelinerun has not been created/running? 

When a status field is set to empty (to resume from pending state), what state is being assigned to `pr.Status`? `started`, `running` ?

Reply:
@pritidesai  if this helps, the webhook will validate before the reconciler does anything

the sequence of events that i understand will be (e.g. imagining you're doing something like `kubectl create/apply -f pipelinerun.yaml` or a tool is doing the equivalent):

1. make call to API server to create PipelineRun
2. API server calls out to any validating/mutating webhooks
3. If the webhooks indicate the PipelineRun is invalid, the creation fails and an error is returned to the user
4. API server stores the pipelinerun in etcd
5. API server notifies any watching connections of the creation of the new pipelinerun
6. pipeline controller is notified of new pipelinerun, then starts reconcling

At (6), the controller would see the "pending" spec.status and would update the conditions to "unknown pipelinerun is pending" 

Later on, you'd actually start the PipelineRun by removing spec.status.pending. At that point this would happen:

1. make call to API server to modify PipelineRun
2. API server calls out to any validating/mutating webhook
3. Removing spec.status.pending is valid, so the mutation is allowed
4. API server stores the pipelinerun in etcd
5. API server notifies any watching connections of the modification of the pipelinerun
6. pipeline controller is notified of modification, then starts reconcling

Then in (6) the controller would see that "pending" had been removed and know to transition from "unknown pipelinerun is pending" to actually starting to execute the pipeline.

After that point, if anyone tried to modify spec.status.pending, the webhook would see in the status of the PipelineRun that it is already running and deny the mutation; the change would never even be stored.

Reply:
thanks a bunch @bobcatfish for the walkthrough 😻 

so between 5 and 6, the process watching these connections of creating a new `pipelineRun` will have to update the conditions to `pending` and looking at the controller logs (also depends on how the monitoring process is designed/implemented and the platform its running on), not much time (in may be part of micro seconds) before a `pipelineRun` actually starts running after getting created. 


> If the PipelineRunPending setting is set after a PipelineRun has been created, the validating webhook should reject the update.

Also, this statement is very confusing, compared to transitioning from 5 to 6, if pipelineRun creation is considered invalid going into pending state, how would the watching process know a `pipelineRun` was initiated?

Overall approach seems reasonable. This will be very interesting to see being implemented 😜 

/lgtm

Reply:
> so between 5 and 6, the process watching these connections of creating a new pipelineRun will have to update the conditions to pending 

Do you mean the new service, e.g. let's call it Concurrency Service, will need to add the "pending" label? You're right that the window to do that in would be impossibly small and racy.

In fact for this to work I think the PipelineRuns would need to be created from the start with the "pending" label. This means that users or tooling would have to include it, and it could be backed up with a validating webhook if desired (i.e. do not allow for creation of PipelineRuns without "spec.status.pending")

Reply:
> Do you mean the new service, e.g. let's call it Concurrency Service, will need to add the "pending" label? You're right that the window to do that in would be impossibly small and racy.

Yup, the Concurrency Service having a very small window. Also, making sure that the Concurrency service add pending label to only the requested runs.

@chhsia0 would like to have details on how `the Concurrency Service` in your environment is making use of this label when we move this TEP to implementable.

Reply:
In our environment, all PipelineRuns will be created in the pending state and activated by the concurrency service when ready. For example, the concurrency service may check how many PipelineRuns are running and start any pending PipelineRuns when there are less PipelineRuns running than some user defined concurrency threshold.

Reply:
> Do you mean the new service, e.g. let's call it Concurrency Service, will need to add the "pending" label? You're right that the window to do that in would be impossibly small and racy.

> Yup, the Concurrency Service having a very small window.

@jbarrick-mesosphere is already saying this but to be totally clear: we **cannot** rely on a watching service to add the pending label (b/c the window would be so small and racy, if its even possible) - using this feature would **require** creating PipelineRuns with spec.status.pending

Reply:
Thanks a bunch @bobcatfish and @jbarrick-mesosphere 

so, run is created in pending mode and watching service is responsible for clearing this label so that the run can be scheduled/executed.

What happens if the pipeline or task definition changes between the time it was sent to pending vs the time it was activated to run? The run continues with the old or new definition? or its discarded saying the definition changed?

When it says the status is set to empty to activate the run, is it the same `status` which was used to send it to `pending`? In that case, isn't it equivalent to creating a new run from the syntax perspective? of course, the controller will check to create a new run or activate the pending run. I am not objecting this but just making it explicit.

Reply:
> What happens if the pipeline or task definition changes between the time it was sent to pending vs the time it was activated to run? The run continues with the old or new definition? or its discarded saying the definition changed?

That's an interesting question - afaik today you can update a PipelineRun in progress and the changes will be picked up and used - e.g. say. you were running a PipelineRun with param X set to "foo", used by Task A, then Task B (which runs after Task A). If, while Task A was running, you changed the value of the param X to "bar", I think Task B would end up using Bar.

I might be wrong but I don't think we have anything in place that would prevent that from happening today, so I think this question applies globally to Run execution and not just to this feature.

> In that case, isn't it equivalent to creating a new run from the syntax perspective?

I think it's very similar to "cancelling" a pipeline run, which is also done by modifying the value of spec.status (i.e. we do not treat this as creating a run, we treat this as a way to communicate the desired state of the existing run)

Reply:
> What happens if the pipeline or task definition changes between the time it was sent to pending vs the time it was activated to run? The run continues with the old or new definition? or its discarded saying the definition changed?

This is a very good question. Depending on what we wanna do, we could prevent that. At the "creation" time and first "reconciliation" loop, we get the `PipelineSpec` (from whatever it comes from) and store it into the status (also known as dereference). We could base the rest of the code and reconciliation loop to look at the status's pipelineSpec (and taskSpec) as source of truth. 

This would fix a bunch of potential problems too (that exists today) : what happens when the def of reference change (also applies to bundles), and it would probably reduce api calls (as we would have all we need at hand).

Reply:
> I might be wrong but I don't think we have anything in place that would prevent that from happening today, so I think this question applies globally to Run execution and not just to this feature.

I agree the specified example is not specific to this feature but applies globally. Yup, we work on modified `spec` so the changes are picked up. I am more concerned with one scenario with this feature is `permanentError` i.e. Task A existed at the time of creating a run in pending mode but at the time executing, if Task A is missing, controller would exit with `missingTaskDef` but the `pr.Status.Spec` would actually have that definition. This is something very rare and applies more to this feature. 

I like what @vdemeester is suggesting to consider `pr.Status.Spec` as a source of truth which is more visible in this feature and `bundles` (hopefully not fetching bundles on every reconciliation loop, though I am not able to recall how its implemented).

> I think it's very similar to "cancelling" a pipeline run, which is also done by modifying the value of spec.status (i.e. we do not treat this as creating a run, we treat this as a way to communicate the desired state of the existing run)

Right, but the existing proposal does not have any special status to cancel pending state and resume executing the run. `spec.Status` is set to `cancel` to cancel existing run,  `spec.Status` is set to `pending` to create run in pending state, spec.Status to empty ``, to resume pending run? 

Also, something more to consider is how long a `pipelineRun` can stay in `pending` state? Controller can create many such runs and forget about them. What happens if those `runs` are never resumed back? This might seem out of scope here but this can fill up `etcd` quickly and can impact the controller.

These are all very important and useful discussions and I appreciate all of your patience 🙏